Написал вход в программу с аргументом командной строки

RatatataBoom opened the pull request.

RatatataBoom committed:
Написал вход в программу с аргументом командной строки

CodingFeles commented on task-4/ShapePacker/ShapePacker/Class Cube.cs:
```diff
@@ -0,0 +1,12 @@
+
+namespace ShapePacker
+{
+    class Class_Cube
```
`class Class_Cube` нет необходимости в названия классов добавлять слово "Class". Это и так будет понятно.

RatatataBoom committed:
[task-4] Дописал программу. Но есть проблемы с командной строкой.

В отладчике программа работает нормально, но не могу запустить с аргументом в виде относительного пути из командной строки.

Tauders commented on task-4/ShapePack/ShapePack/Container.cs:
```diff
@@ -0,0 +1,26 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
```
Мелочь: нужны ли столько namespace'ов?

Tauders commented on task-4/ShapePack/ShapePack/Container.cs:
```diff
@@ -0,0 +1,26 @@
+{
+    class Container
+    {
+        public double Volume()
```
Код стиль: имя функции должно отвечать на вопрос глагола, то есть `GetVolume`, а ещё лучше вложить в название суть того, что в методе происходит, даже если это будет удлинять название метода, т.е. `GetVolumeFromConsoleInput` или что-то такое тоже допустимо.

Tauders commented on task-4/ShapePack/ShapePack/Container.cs:
```diff
@@ -0,0 +1,26 @@
+            catch (Exception)
+            {
+                //Console.Write("Вы ввели недопустимый символ. Нажмите Enter.\n");
+                return 0;
```
Надо ли отлавливать такого рода исключения и возвращать 0? Как различать случай, когда пользователь ввёл 0 или когда вылетела ошибка?

Tauders commented on task-4/ShapePack/ShapePack/Container.cs:
```diff
@@ -0,0 +1,26 @@
+            try
+            {
+                Console.Write("Введите объем контейнера(объём контейнера вводится как вещественное число):");
+                double volume = double.Parse(Console.ReadLine());
```
Что должно происходить, если пользователь введёт отрицательное число?

Tauders commented on task-4/ShapePack/ShapePack/Container.cs:
```diff
@@ -0,0 +1,26 @@
+        {
+            try
+            {
+                Console.Write("Введите объем контейнера(объём контейнера вводится как вещественное число):");
```
Опечатка: между текстом и "(" надо пробел добавить

Tauders commented on task-4/ShapePack/ShapePack/Container.cs:
```diff
@@ -0,0 +1,26 @@
+            }
+            catch (Exception)
+            {
+                //Console.Write("Вы ввели недопустимый символ. Нажмите Enter.\n");
```
Код стиль: лишние комментарии надо убирать

Tauders commented on task-4/ShapePack/ShapePack/Container.cs:
```diff
@@ -0,0 +1,26 @@
+            try
+            {
+                Console.Write("Введите объем контейнера(объём контейнера вводится как вещественное число):");
+                double volume = double.Parse(Console.ReadLine());
+                return volume;
```
Можно написать 
```c#
return double.Parse(Console.ReadLine());
```

Но это частично вкусовщина, пока правил на проекте нет, так что если нравится текущий вариант, то оставляй его.

Tauders commented on task-4/ShapePack/ShapePack/Cube.cs:
```diff
@@ -0,0 +1,38 @@
+            this.name = name;
+            this.height = height;
+        }
+        private string name
+        {
+            get; set;
+        }
+        private double height
+        {
+            get; set;
+        }
```
Нужны ли тебе эти приватные свойства? (https://docs.microsoft.com/ru-ru/dotnet/csharp/programming-guide/classes-and-structs/properties)
Не проще ли сделать поля (https://docs.microsoft.com/ru-ru/dotnet/csharp/programming-guide/classes-and-structs/fields) приватные?

Код стиль: приватные свойства имеют другой стиль именования.

Tauders commented on task-4/ShapePack/ShapePack/Cube.cs:
```diff
@@ -0,0 +1,38 @@
+        {
+            get; set;
+        }
+        private double Write()
```
Нужна ли эта функция?

Код стиль: название.

Tauders commented on task-4/ShapePack/ShapePack/Cube.cs:
```diff
@@ -0,0 +1,38 @@
+            double height = double.Parse(Console.ReadLine());
+            return height;
+        }
+        public double Volume()
```
Код стиль: название.

Tauders commented on task-4/ShapePack/ShapePack/Cylinder.cs:
```diff
@@ -0,0 +1,44 @@
+            this.radius = radius;
+            this.height = height;
+        }
+        private string name
+        {
+            get; set;
+        }
+        private double height
+        {
+            get; set;
+        }
+        private double radius
+        {
+            get; set;
+        }
```
Код стиль: название.

Нужны ли вообще свойства и не хватит ли полей?

Tauders commented on task-4/ShapePack/ShapePack/Cylinder.cs:
```diff
@@ -0,0 +1,44 @@
+        }
+        private double Write()
+        {
+            Console.Write("Введите Высоту:");
+            double height = double.Parse(Console.ReadLine());
+            return height;
+            Console.Write("Введите Радиус:");
+            double radius = double.Parse(Console.ReadLine());
+            return radius;
```
Несколько `return` в коде не имеет смысла, код выполняется сверху вниз и если наткнётся на первый return, то весь дальнейший код не будет выполняться.

Tauders commented on task-4/ShapePack/ShapePack/Cylinder.cs:
```diff
@@ -0,0 +1,44 @@
+        {
+            get; set;
+        }
+        private double Write()
```
Нужна ли эта функция? Что она должна была делать?

Код стиль: название.

Tauders commented on task-4/ShapePack/ShapePack/Cylinder.cs:
```diff
@@ -0,0 +1,44 @@
+            double radius = double.Parse(Console.ReadLine());
+            return radius;
+        }
+        public double Volume()
```
Код стиль: название.

Tauders commented on task-4/ShapePack/ShapePack/Program.cs:
```diff
@@ -0,0 +1,168 @@
+{
+    class Program
+    {
+        static void Insert(ref string[] array, string value, int index)
```
Зачем эта функция в классе `Program`, может её стоит вынести? Нужна ли она вообще? 

Параметры на вход: Зачем здесь нужны `ref`? Не проще ли всегда возвращать массив новый? Зачем `index`, если по использованию он всегда 0, может проще его не передавать и использовать всегда 0?

Код стиль: название, по нему вообще не понятно, что происходит.

Tauders commented on task-4/ShapePack/ShapePack/Program.cs:
```diff
@@ -0,0 +1,168 @@
+
+            newArray[index] = value;
+
+            for (int i = 0; i < index; i++)
+            {
+                newArray[i] = array[i];
+            }
```
Так как `index` всегда равен 0, то цикл не выполняется при текущем условии.

0 итерация - (i == 0) < 0 и цикл заканчивается

Tauders commented on task-4/ShapePack/ShapePack/Program.cs:
```diff
@@ -0,0 +1,168 @@
+            {
+                newArray[i] = array[i];
+            }
+            for (int i = index; i < array.Length; i++)
+            {
+                newArray[i + 1] = array[i];
+            }
```
Если надо переместить один массив в другой, то стоит задаться вопросом - первый ли ты человек, которому такое надо? 

Если нет, то стоит попробовать найти готовое решение в мощностях языка, библиотеках, которые уже используешь, новых для проекта библиотеках, решениях в интернете, или написать самому.

В этом случае может подойти `Array.Copy`

Tauders commented on task-4/ShapePack/ShapePack/Program.cs:
```diff
@@ -0,0 +1,168 @@
+                // Проверка на наличие аргумента
+                if (args.Length == 0)
+                {
+                    Console.WriteLine("Требуется ввести путь до файла с исходными данными фигур");
```
Мелочь: немного вводящее в заблуждение сообщение - непонятно, куда ввести путь, я бы ожидал в консоль, а тут не тот случай.

Tauders commented on task-4/ShapePack/ShapePack/Program.cs:
```diff
@@ -0,0 +1,168 @@
+                    continue;
+                }
+
+                string pathToFile = System.IO.Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "input.txt");
```
Нужна ли эта переменная? Нужен ли такой длинный путь `System.IO.Path.Combine`? Почему не хватит просто `Path.Combine`?

Tauders commented on task-4/ShapePack/ShapePack/Program.cs:
```diff
@@ -0,0 +1,168 @@
+                string pathToFile = System.IO.Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "input.txt");
+
+
+                string[] Shapes = new string[0];
```
Код стиль: название.

У переменных внутри функции они с маленькой буквы, так что просто `shapes`.

По поводу выделения массива пустого, сама Visual Studio советует это правило, и как такого поведения избегать: https://docs.microsoft.com/ru-ru/dotnet/fundamentals/code-analysis/quality-rules/ca1825

Tauders commented on task-4/ShapePack/ShapePack/Program.cs:
```diff
@@ -0,0 +1,168 @@
+                // Читаю файл -> Разбиваю на строки -> Строки разбиваю на слова
+                try
+                {
+                    string[] readText = File.ReadAllLines(args[0]);
```
Код стиль: название.

Переменная это существительное, так что глагол `readText` не очень подходит. `linesOfFile` или `fileLines` или `stringsOfFile` лучше подойдёт.

А если первым аргументом будет не путь до файла? Что произойдёт?

Зачем ты считываешь файл здесь, если дальше, при условии ввода неправильного объёма контейнера ты не будешь использовать эти данные?

Tauders commented on task-4/ShapePack/ShapePack/Program.cs:
```diff
@@ -0,0 +1,168 @@
+                {
+                    string[] readText = File.ReadAllLines(args[0]);
+                    
+                    var Container = new Container();
```
Код стиль: название.

Tauders commented on task-4/ShapePack/ShapePack/Program.cs:
```diff
@@ -0,0 +1,168 @@
+                    string[] readText = File.ReadAllLines(args[0]);
+                    
+                    var Container = new Container();
+                    double containerVolume = Container.Volume();
+                    if (containerVolume == 0)
```
А если отрицательное число?

Tauders replied:
Дико не очевидно, что внутри `Container.Volume()` происходит запрос с консоли. Почему бы не вынести чтение с консоли сюда и обернуть в ещё один `try {} catch {}`?

Это работает, если будут вопросы, пиши, я отпишу, как.

Tauders commented on task-4/ShapePack/ShapePack/Program.cs:
```diff
@@ -0,0 +1,168 @@
+                    {
+                        string[] subString = line.Split(' ');
+
+                        for (int i = 0; i < subString.Length; i++){}
```
Пустой цикл, который ничего не делает.

Tauders commented on task-4/ShapePack/ShapePack/Program.cs:
```diff
@@ -0,0 +1,168 @@
+                        double height, radius, sideA, sideB;
+                        string name;
+                        
+                        //Логика с фигурами и контейнером. По итогу имя фигуры добалвляется в массив строк.
```
Опечатка: "добалвляется"

Tauders commented on task-4/ShapePack/ShapePack/Program.cs:
```diff
@@ -0,0 +1,168 @@
+                        //Логика с фигурами и контейнером. По итогу имя фигуры добалвляется в массив строк.
+                        while (containerVolume >= 0)
+                        {
+                            if (subString[0] == "Cube")
```
То, как сделано здесь, абсолютно нормальное решение, в плане использования `if else`, но ради дополнительной тренировки дополнительное задание:

Добавь Enum с всеми поддерживаемыми видами фигур (https://docs.microsoft.com/ru-ru/dotnet/csharp/language-reference/builtin-types/enum)

Сначала преобразуй первый элемент массива в Enum, а потом на основе его сделай `switch case`.

Tauders commented on task-4/ShapePack/ShapePack/Program.cs:
```diff
@@ -0,0 +1,168 @@
+                    
+                    foreach (string line in readText)
+                    {
+                        string[] subString = line.Split(' ');
```
Что будет, если строка будет пустая или содержать некорректные символы?

Tauders commented on task-4/ShapePack/ShapePack/Program.cs:
```diff
@@ -0,0 +1,168 @@
+                            {
+                                name = subString[1];
+                                height = double.Parse(subString[2]);
+                                var Cube = new Cube(height, name);
```
Код стиль: название.

RatatataBoom replied:
не понял что не так с названиями

Tauders replied:
`var Cube` - почему название `Cube` с большой буквы? Это же локальная переменная.

Tauders commented on task-4/ShapePack/ShapePack/Program.cs:
```diff
@@ -0,0 +1,168 @@
+                        {
+                            if (subString[0] == "Cube")
+                            {
+                                name = subString[1];
+                                height = double.Parse(subString[2]);
```
Зачем эти переменные объявлены ранее, если используются только внутри каждой ветки условия и не используются где-то обще?

Опять же, что будет, если переменные не правильно парсятся или их неправильное количество?

Tauders commented on task-4/ShapePack/ShapePack/Program.cs:
```diff
@@ -0,0 +1,168 @@
+                            }
+                            else if (subString[0] == "Sphere")
+                            {
+                                name = subString[1];
+                                radius = double.Parse(subString[2]);
```
Такие же замечания, как у куба.

Tauders commented on task-4/ShapePack/ShapePack/Program.cs:
```diff
@@ -0,0 +1,168 @@
+                            }
+                            else if (subString[0] == "Pyramid")
+                            {
+                                name = subString[1];
+                                height = double.Parse(subString[2]);
+                                sideA = double.Parse(subString[3]);
+                                sideB = double.Parse(subString[4]);
```
Такие же замечания, как у куба.

Tauders commented on task-4/ShapePack/ShapePack/Program.cs:
```diff
@@ -0,0 +1,168 @@
+                            }
+                            else if (subString[0] == "Cylinder")
+                            {
+                                name = subString[1];
+                                height = double.Parse(subString[2]);
+                                radius = double.Parse(subString[3]);
```
Такие же замечания, как у куба.

Tauders commented on task-4/ShapePack/ShapePack/Program.cs:
```diff
@@ -0,0 +1,168 @@
+                                    Insert(ref Shapes, name, 0);
+                                }
+                            }
+                            break;
```
Зачем здесь break?

RatatataBoom replied:
у меня без этого break программа не идет дальше ввода объема контейнера

Tauders replied:
Я забыл указать ещё один важный момент, из-за которого брейк и нужен. Что делает `break` в целом? Прерывает цикл (и не только, но в этом случае упростим). 

Почему тебе надо прерывать цикл каждую итерацию цикла? Имеет ли смысл делать весь этот блок кода в виде цикла?

Tauders replied:
Дальше я пишу 
> Зачем здесь цикл?

Вот то, что сверху это подводка к более развёрнотому ответу на вопрос.

Tauders commented on task-4/ShapePack/ShapePack/Program.cs:
```diff
@@ -0,0 +1,168 @@
+                        string name;
+                        
+                        //Логика с фигурами и контейнером. По итогу имя фигуры добалвляется в массив строк.
+                        while (containerVolume >= 0)
```
Зачем здесь цикл?

Tauders commented on task-4/ShapePack/ShapePack/Program.cs:
```diff
@@ -0,0 +1,168 @@
+                }
+                catch (Exception)
+                {
+                    Console.WriteLine("Не верно введен путь, проверьте правильность");
```
Супер общая ошибка, хотя здесь может быть куча причин ошибки - неправильно строка распарсилась, неправильный формат, недостаточно данных, и т.д. 

Разбей на более точечные причины, чтобы было понятно, что пошло не так.

Tauders commented on task-4/ShapePack/ShapePack/Program.cs:
```diff
@@ -0,0 +1,168 @@
+                    Console.WriteLine("Не верно введен путь, проверьте правильность");
+                    return;
+                }
+                int quantiyShapes = Shapes.Length;
```
Опечатка: "quantiyShapes"

Tauders commented on task-4/ShapePack/ShapePack/Program.cs:
```diff
@@ -0,0 +1,168 @@
+                    return;
+                }
+                int quantiyShapes = Shapes.Length;
+                string quantiyShapesString = Convert.ToString(quantiyShapes);
```
Почему бы просто не вызвать `quantiyShapes.ToString()` или даже `Shapes.Length.ToString()`?

Tauders commented on task-4/ShapePack/ShapePack/Program.cs:
```diff
@@ -0,0 +1,168 @@
+
+                Insert(ref Shapes, quantiyShapesString, 0);
+
+                Console.Write("Если хотите сохранить результат в файл? нажмите 'F'\nЕсли хотите вывести результат на экран? нажмите 'M'");
```
Мелочь: Странно, почему там вопрос после "файл" и "экран"?

Tauders commented on task-4/ShapePack/ShapePack/Program.cs:
```diff
@@ -0,0 +1,168 @@
+
+                Console.Write("Если хотите сохранить результат в файл? нажмите 'F'\nЕсли хотите вывести результат на экран? нажмите 'M'");
+
+                ConsoleKeyInfo chinput = Console.ReadKey();
```
Код стиль: непонятное сокращение.

Tauders commented on task-4/ShapePack/ShapePack/Program.cs:
```diff
@@ -0,0 +1,168 @@
+                    case ConsoleKey.F:
+                        {
+                            Console.Clear();
+                            File.WriteAllLines(@"C:\Users\anato\source\repos\ShapePack\output.txt", Shapes);
```
Надо переделать на относительный путь, но пока забьём, просто держать в голове.

Tauders commented on task-4/ShapePack/ShapePack/Program.cs:
```diff
@@ -0,0 +1,168 @@
+                            
+                        }
+                    default:
+                        Console.WriteLine("Unknown Command.");
```
А почему на английском? :)

Tauders commented on task-4/ShapePack/ShapePack/Properties/launchSettings.json:
```diff
@@ -0,0 +1,9 @@
+    "ShapePack": {
+      "commandName": "Project",
+      "commandLineArgs": "input.txt",
+      "workingDirectory": "C:\\Users\\anato\\source\\repos\\ShapePack\\ShapePack\\bin\\Debug\\net5.0"
```
А если убрать это, не будет ли `input.txt` просто браться относительно? Посмотри на свойства файла в Visual Studio, то есть добавь его сначала в проект через `Add -> existing` или как-то так, и потом посмотри, как заставить его копироваться в папку, куда собирается твоё приложение ( в  `C:\\Users\\anato\\source\\repos\\ShapePack\\ShapePack\\bin\\Debug\\net5.0`)

Подсказка: это связано с режимами копирования.

Tauders commented on task-4/ShapePack/ShapePack/Pyramid.cs:
```diff
@@ -0,0 +1,52 @@
+    {
+        public Pyramid(double height, double sideA, double sideB, string name)
+        {
+            this.name = name;
+            this.height = height;
+            this.sideA = sideA;
+            this.sideB = sideB;
```
Описано ранее, для других фигур.

Tauders commented on task-4/ShapePack/ShapePack/Sphere.cs:
```diff
@@ -0,0 +1,36 @@
+    {
+        public Sphere(double radius, string name)
+        {
+            this.name = name;
+            this.radius = radius;
```
Для других фигур.

Tauders commented on task-4/ShapePack/input old.txt:
```diff
@@ -0,0 +1,5 @@
+Cube Шестигранник 2
+Sphere ВрекингБолл 100
+Cube КубикЛьда 3
+Pyramid ХанойскаяПирамидка 4 4 2
+Cylinder Шляпа 29 30
```
Нужен этот файл?

Tauders reviewed (changes requested):
Мелкие замечания:
Удали файл `ShapePack.exe` из корня папки.
Организуй одинаковые классы по смыслу в одну папку, например все фигуры в папку фигур.

Крупные замечания/предложения:
Эту задачу можно решить через большее использование ООП. Попробуй следующее:
Сделай общий абстрактный класс `Shape`, в котором будут супер общие вещи, например имя и метод для вычисления объёма.

Пусть каждый класс фигуры наследует этот класс и реализует метод вычисления объёма по своему.

Что это даст?

В файле `Program.cs` ты сможешь сказать, что у тебя есть некоторые абстрактные фигуры (их массив или список), которые ты будешь получать из файла. Потом ты возьмёшь и заполнишь конкретными фигурами свой массив абстрактных фигур (сейчас это происходит в Program.cs в `if else`). 
После этого у тебя будет набор фигур, которые знают об абстрактному "себе" 2 вещи - название и объём, как раз это тебе и надо, по сути.

Дальше, имея этот массив, ты сможешь проще и понятнее обработать ситуацию с контейнером.

Могу расписать подробнее, но попробуй на основе этой подсказки сделать решение.
